--- nginx/auto/make.orig	2015-05-08 13:46:12.000000000 +0100
+++ nginx/auto/make	2015-05-08 14:35:23.000000000 +0100
@@ -24,6 +24,8 @@
 CFLAGS = $CFLAGS
 CPP =	$CPP
 LINK =	$LINK
+CXXFLAGS = $CXXFLAGS
+CXX = $CXX
 
 END
 
@@ -407,10 +409,19 @@
 
 if test -n "$NGX_ADDON_SRCS"; then
 
+	echo " "
+	echo "--- Addon Modules ---"
+	echo " "
+
     ngx_cc="\$(CC) $ngx_compile_opt \$(CFLAGS) $ngx_use_pch \$(ALL_INCS)"
 
+	ngx_cxx="\$(CXX) -c \$(CXXFLAGS) \$(ALL_INCS)"
+
     for ngx_src in $NGX_ADDON_SRCS
     do
+
+    	echo $(basename $ngx_src)
+
         ngx_obj="addon/`basename \`dirname $ngx_src\``"
 
         ngx_obj=`echo $ngx_obj/\`basename $ngx_src\` \
@@ -424,14 +435,33 @@
 
         ngx_src=`echo $ngx_src | sed -e "s/\//$ngx_regex_dirsep/g"`
 
+		if [[ "$ngx_src" == *.cc ]] ; then
+
         cat << END                                            >> $NGX_MAKEFILE
 
 $ngx_obj:	\$(ADDON_DEPS)$ngx_cont$ngx_src
+	$ngx_cxx$ngx_tab$ngx_objout$ngx_obj$ngx_tab$ngx_src$NGX_AUX
+
+END
+
+		else
+
+
+        cat << END                                            >> $NGX_MAKEFILE
+
+
+$ngx_obj:	\$(ADDON_DEPS)$ngx_cont$ngx_src
 	$ngx_cc$ngx_tab$ngx_objout$ngx_obj$ngx_tab$ngx_src$NGX_AUX
 
 END
+
+    	fi
+
      done
 
+	echo " "
+	echo "------------"
+
 fi
 
 
